server/base/src/main/java/org/apache/accumulo/server/metrics/MetricsInfoImpl.java

```diff
@@ -0,0 +1,340 @@
+  }
+
+  @Override
+  public void addMetricsProducers(MetricsProducer... producer) {
+    lock.lock();
+    try {
+      if (composite == null) {
+        producers.addAll(Arrays.asList(producer));
+      } else {
+        Arrays.stream(producer).forEach(p -> p.registerMetrics(composite));
+      }
+    } finally {
+      lock.unlock();
+    }
+  }
+
+  @Override
+  public MeterRegistry getRegistry() {
+    lock.lock();
+    try {
+      if (composite == null) {
+        throw new IllegalStateException("metrics have not been initialized, call init() first");
+      }
+    } finally {
+      lock.unlock();
+    }
+    return composite;
+  }
+
+  @Override
+  public void init() {
+    lock.lock();
+    try {
+      if (composite != null) {
+        LOG.warn("metrics registry has already been initialized");
+        return;
+      }
+      composite = new CompositeMeterRegistry();
+      composite.config().commonTags(commonTags.values());
+
+      LOG.info("Metrics initialization. common tags: {}", commonTags);
+
+      boolean jvmMetricsEnabled =
+          context.getConfiguration().getBoolean(Property.GENERAL_MICROMETER_JVM_METRICS_ENABLED);
+
+      if (jvmMetricsEnabled) {
+        LOG.info("enabling detailed jvm, classloader, jvm gc and process metrics");
+        new ClassLoaderMetrics().bindTo(composite);
+        new JvmMemoryMetrics().bindTo(composite);
+        jvmGcMetrics = new JvmGcMetrics();
+        jvmGcMetrics.bindTo(composite);
+        new ProcessorMetrics().bindTo(composite);
+        new JvmThreadMetrics().bindTo(composite);
+      }
+
+      MeterFilter replicationFilter = new MeterFilter() {
+        @Override
+        public DistributionStatisticConfig configure(Meter.Id id,
+            @NonNull DistributionStatisticConfig config) {
+          if (id.getName().equals("replicationQueue")) {
+            return DistributionStatisticConfig.builder().percentiles(0.5, 0.75, 0.9, 0.95, 0.99)
+                .expiry(Duration.ofMinutes(10)).build().merge(config);
+          }
+          return config;
+        }
+      };
+
+      if (metricsEnabled()) {
+        // user specified registries
+        String userRegistryFactories =
+            context.getConfiguration().get(Property.GENERAL_MICROMETER_FACTORY);
+
+        for (String factoryName : getTrimmedStrings(userRegistryFactories)) {
+          try {
+            MeterRegistry registry = getRegistryFromFactory(factoryName, context);
+            registry.config().commonTags(commonTags.values());
+            registry.config().meterFilter(replicationFilter);
+            addRegistry(registry);
+          } catch (ClassNotFoundException | NoSuchMethodException | InvocationTargetException
+              | InstantiationException | IllegalAccessException ex) {
+            LOG.warn("Could not load registry {}", factoryName, ex);
+          }
+        }
+      }
+
+      Metrics.globalRegistry.add(composite);
+      pendingRegistries.forEach(registry -> composite.add(registry));
```

Comment:
is there a reason to add the registries to the composite *after* setting the composite on the global registry? I'm wondering if the composite should be constructed first and then added to the global registry.

Reply:
Addressed in 601f167e80